### DISCUSSION TITLE
Is all API thread-safe? (e.g. create two threads to invoke model.transcribe for two files)

### DISCUSSION THREAD

**Discussion by grimmerk** [openai/whisper#296] [Q&A] [answered]
I saw we can use multi-thread to invoke APIs, [ref](https://github.com/openai/whisper/pull/278). But the question is not the case. 

I also see Pytorch's inference is thread safe, [ref](https://discuss.pytorch.org/t/is-inference-thread-safe/88583). However, I'd like to double-confirm with you.

**Comment by jongwook** [comment, marked as answer]
A short answer is no.

The KV caching mechanism uses forward hooks which are installed in the module objects and will cause race issues when used by multiple threads. The `--threads` option provides a more low-level control on how the CPU operations are parallelized, but it's less relevant if you're using GPU.

Even if you disabled KV caching, multi-threaded usage will be generally inefficient because of the GIL. Multiprocessing will buy you more, like in [Multi-Instance GPU](https://www.nvidia.com/en-us/technologies/multi-instance-gpu/), and it may be more flexible for multithreading if you use the PyTorch C++ API as in your second reference.

If you're integrating Whisper with a serving layer, it may support automatic batching [e.g. in TensorFlow Serving](https://github.com/tensorflow/serving/blob/master/tensorflow_serving/batching/README.md), which is usually more efficient use of GPU resource.

**Comment by silvacarl2** [comment]
what is the command line or switches to enable whisper to take advnateg of multiple GPUs?
